File: server/control-api/internal/knowledge/localpg/chunk.go
package localpg

import (
	"regexp"
	"strings"
	"unicode/utf8"
)

const (
	maxChunkRunes     = 2000
	hardSplitLookback = 400
	maxChunks         = 500
)

var (
	phonePattern    = regexp.MustCompile(`(?:\+?86[- ]?)?1[3-9]\d{9}`)
	emailPattern    = regexp.MustCompile(`[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}`)
	blankParagraphs = regexp.MustCompile(`\n\s*\n+`)
	dateLine        = regexp.MustCompile(`(?i)^\d{4}([./\-年])\d{1,2}月?(?:\s*[-–—至到]+\s*(?:\d{4}([./\-年])\d{1,2}月?|至今|现在|present))?`)
	bulletLine      = regexp.MustCompile(`^([•·●○■\-*、]|（\d+）|\(\d+\)|\d+[\.、])\s*`)
)

var sectionHeadings = []string{
	"工作经历", "工作经验", "实习经历", "项目经历", "项目经验",
	"教育背景", "教育经历", "专业技能", "技能特长", "自我评价", "个人总结",
	"experience", "work experience", "projects", "education", "skills", "summary",
}

type Chunk struct {
	Index   int
	Content string
}

func ChunkResume(candidateName, text string) []Chunk {
	cleaned := cleanText(text)
	if strings.TrimSpace(cleaned) == "" {
		return nil
	}
	sections := splitSections(cleaned)
	var raw []string
	for _, section := range sections {
		raw = append(raw, splitSection(candidateName, section.heading, section.body)...)
	}
	if len(raw) == 0 {
		raw = splitOversized(cleaned)
	}
	chunks := make([]Chunk, 0, len(raw))
	for _, content := range raw {
		content = strings.TrimSpace(content)
		if content == "" {
			continue
		}
		chunks = append(chunks, Chunk{Index: len(chunks), Content: content})
		if len(chunks) >= maxChunks {
			break
		}
	}
	return chunks
}

type section struct {
	heading string
	body    string
}

func cleanText(text string) string {
	text = strings.ReplaceAll(text, "\r\n", "\n")
	text = strings.ReplaceAll(text, "\r", "\n")
	text = phonePattern.ReplaceAllString(text, "")
	text = emailPattern.ReplaceAllString(text, "")
	lines := strings.Split(text, "\n")
	kept := make([]string, 0, len(lines))
	for _, line := range lines {
		line = strings.Join(strings.Fields(line), " ")
		if utf8.RuneCountInString(line) < 2 {
			if strings.TrimSpace(line) == "" {
				kept = append(kept, "")
			}
			continue
		}
		kept = append(kept, line)
	}
	return strings.TrimSpace(strings.Join(kept, "\n"))
}

func splitSections(text string) []section {
	lines := strings.Split(text, "\n")
	var sections []section
	current := section{heading: "", body: ""}
	flush := func() {
		body := strings.TrimSpace(current.body)
		if body == "" {
			return
		}
		sections = append(sections, section{heading: current.heading, body: body})
	}
	for _, line := range lines {
		if heading, ok := matchHeading(line); ok {
			flush()
			current = section{heading: heading, body: ""}
			continue
		}
		if current.body != "" {
			current.body += "\n"
		}
		current.body += line
	}
	flush()
	if len(sections) == 0 && strings.TrimSpace(text) != "" {
		return []section{{heading: "", body: strings.TrimSpace(text)}}
	}
	return sections
}

func matchHeading(line string) (string, bool) {
	trimmed := strings.TrimSpace(line)
	trimmed = strings.Trim(trimmed, "：:")
	normalized := strings.ToLower(trimmed)
	for _, heading := range sectionHeadings {
		if normalized == strings.ToLower(heading) {
			return heading, true
		}
	}
	return "", false
}

func splitSection(candidate, heading, body string) []string {
	if isExperienceHeading(heading) {
		return chunkBlocks(candidate, heading, splitExperienceBlocks(body))
	}
	if heading == "" {
		blocks := splitBlankParagraphs(body)
		if len(blocks) <= 1 {
			items := splitBulletItems(body)
			if len(items) > 1 {
				blocks = items
			}
		}
		return chunkBlocks(candidate, heading, blocks)
	}
	title := firstLineTitle(body)
	prefixed := withPrefix(candidate, heading, title, body)
	return splitOversized(prefixed)
}

func chunkBlocks(candidate, heading string, blocks []string) []string {
	var out []string
	for _, block := range blocks {
		title := firstLineTitle(block)
		prefixed := withPrefix(candidate, heading, title, block)
		out = append(out, splitOversized(prefixed)...)
	}
	return out
}

func splitBlankParagraphs(body string) []string {
	parts := blankParagraphs.Split(body, -1)
	var blocks []string
	for _, part := range parts {
		part = strings.TrimSpace(part)
		if part != "" {
			blocks = append(blocks, part)
		}
	}
	if len(blocks) == 0 && strings.TrimSpace(body) != "" {
		return []string{strings.TrimSpace(body)}
	}
	return blocks
}

func isExperienceHeading(heading string) bool {
	switch strings.ToLower(heading) {
	case "工作经历", "工作经验", "实习经历", "项目经历", "项目经验", "experience", "work experience", "projects":
		return true
	default:
		return false
	}
}

func splitExperienceBlocks(body string) []string {
	lines := strings.Split(body, "\n")
	var blocks []string
	var current []string
	flush := func() {
		text := strings.TrimSpace(strings.Join(current, "\n"))
		if text != "" {
			blocks = append(blocks, text)
		}
		current = nil
	}
	for _, line := range lines {
		if dateLine.MatchString(strings.TrimSpace(line)) && len(current) > 0 {
			flush()
		}
		current = append(current, line)
	}
	flush()
	if len(blocks) == 0 && strings.TrimSpace(body) != "" {
		return []string{strings.TrimSpace(body)}
	}
	return blocks
}

func firstLineTitle(block string) string {
	line := strings.TrimSpace(strings.Split(block, "\n")[0])
	if utf8.RuneCountInString(line) > 80 {
		return string([]rune(line)[:80])
	}
	return line
}

func withPrefix(candidate, heading, title, body string) string {
	label := strings.TrimSpace(strings.Join([]string{candidate, heading, title}, " | "))
	label = strings.Trim(label, " |")
	if label == "" {
		return strings.TrimSpace(body)
	}
	return "[" + label + "]\n" + strings.TrimSpace(body)
}

func splitOversized(text string) []string {
	if utf8.RuneCountInString(text) <= maxChunkRunes {
		if strings.TrimSpace(text) == "" {
			return nil
		}
		return []string{text}
	}
	body := text
	prefix := ""
	if strings.HasPrefix(text, "[") {
		if idx := strings.Index(text, "]\n"); idx >= 0 {
			prefix = text[:idx+2]
			body = text[idx+2:]
		}
	}
	items := splitBulletItems(body)
	if len(items) > 1 {
		var out []string
		for _, item := range items {
			out = append(out, splitOversized(strings.TrimSpace(prefix+item))...)
		}
		return out
	}
	return hardSplit(text, prefix)
}

func splitBulletItems(body string) []string {
	lines := strings.Split(body, "\n")
	var items []string
	var current []string
	flush := func() {
		text := strings.TrimSpace(strings.Join(current, "\n"))
		if text != "" {
			items = append(items, text)
		}
		current = nil
	}
	found := false
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if bulletLine.MatchString(trimmed) && len(current) > 0 {
			found = true
			flush()
		}
		current = append(current, line)
	}
	flush()
	if !found {
		return nil
	}
	return items
}

func hardSplit(text, prefix string) []string {
	runes := []rune(text)
	var out []string
	start := 0
	for start < len(runes) {
		remaining := runes[start:]
		if len(remaining) <= maxChunkRunes {
			piece := strings.TrimSpace(string(remaining))
			if start > 0 && prefix != "" && !strings.HasPrefix(piece, "[") {
				piece = strings.TrimSpace(prefix + piece)
			}
			if piece != "" {
				out = append(out, piece)
			}
			break
		}
		window := remaining
		if len(window) > maxChunkRunes {
			window = remaining[:maxChunkRunes]
		}
		cut := findSplit(window)
		if cut <= 0 {
			cut = maxChunkRunes
		}
		piece := strings.TrimSpace(string(remaining[:cut]))
		if start > 0 && prefix != "" && !strings.HasPrefix(piece, "[") {
			piece = strings.TrimSpace(prefix + piece)
		}
		if piece != "" {
			out = append(out, piece)
		}
		start += cut
	}
	return out
}

func findSplit(window []rune) int {
	if len(window) == 0 {
		return 0
	}
	begin := 0
	if len(window) > hardSplitLookback {
		begin = len(window) - hardSplitLookback
	}
	for i := len(window) - 1; i >= begin; i-- {
		if window[i] == '\n' {
			if i+1 < len(window) && window[i+1] == '\n' {
				return i + 2
			}
			return i + 1
		}
	}
	for i := len(window) - 1; i >= begin; i-- {
		if window[i] == '。' || window[i] == '！' || window[i] == '？' {
			if i > 0 && window[i-1] == window[i] {
				continue
			}
			return i + 1
		}
	}
	return 0
}

File: src-tauri/src/materials/chunk.rs
pub const CHUNKER_VERSION: &str = "resume-semantic-v1";

const MAX_CHUNK_RUNES: usize = 2000;
const HARD_SPLIT_LOOKBACK: usize = 400;
const MAX_CHUNKS: usize = 500;

const SECTION_HEADINGS: &[&str] = &[
    "工作经历",
    "工作经验",
    "实习经历",
    "项目经历",
    "项目经验",
    "教育背景",
    "教育经历",
    "专业技能",
    "技能特长",
    "自我评价",
    "个人总结",
    "experience",
    "work experience",
    "projects",
    "education",
    "skills",
    "summary",
];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MaterialChunk {
    pub index: i64,
    pub content: String,
    pub section: String,
    pub start_char: usize,
    pub end_char: usize,
    pub size_estimate: i64,
}

struct Section {
    heading: String,
    body: String,
}

pub fn chunk_text(source_label: &str, text: &str) -> Vec<MaterialChunk> {
    let cleaned = clean_text(text);
    if cleaned.trim().is_empty() {
        return Vec::new();
    }

    let mut raw = Vec::new();
    let mut sections_for_raw = Vec::new();
    for section in split_sections(&cleaned) {
        let pieces = split_section(source_label, &section.heading, &section.body);
        sections_for_raw.extend(std::iter::repeat_n(section.heading.clone(), pieces.len()));
        raw.extend(pieces);
    }
    if raw.is_empty() {
        raw = split_oversized(&cleaned);
        sections_for_raw = vec![String::new(); raw.len()];
    }

    let mut chunks = Vec::new();
    for (content, section) in raw.into_iter().zip(sections_for_raw) {
        let content = content.trim().to_owned();
        if content.is_empty() {
            continue;
        }
        let (start_char, end_char) = locate_offsets(&cleaned, &content);
        chunks.push(MaterialChunk {
            index: chunks.len() as i64,
            size_estimate: content.chars().count() as i64,
            start_char,
            end_char,
            section,
            content,
        });
        if chunks.len() >= MAX_CHUNKS {
            break;
        }
    }
    chunks
}

fn clean_text(text: &str) -> String {
    let text = text.replace("\r\n", "\n").replace('\r', "\n");
    let mut kept = Vec::new();
    for line in text.lines() {
        let line = line.split_whitespace().collect::<Vec<_>>().join(" ");
        if line.chars().count() < 2 {
            if line.trim().is_empty() {
                kept.push(String::new());
            }
            continue;
        }
        kept.push(line);
    }
    kept.join("\n").trim().to_owned()
}

fn split_sections(text: &str) -> Vec<Section> {
    let mut sections = Vec::new();
    let mut current = Section {
        heading: String::new(),
        body: String::new(),
    };
    let flush = |current: &mut Section, sections: &mut Vec<Section>| {
        let body = current.body.trim();
        if body.is_empty() {
            return;
        }
        sections.push(Section {
            heading: current.heading.clone(),
            body: body.to_owned(),
        });
    };

    for line in text.lines() {
        if let Some(heading) = match_heading(line) {
            flush(&mut current, &mut sections);
            current = Section {
                heading,
                body: String::new(),
            };
            continue;
        }
        if !current.body.is_empty() {
            current.body.push('\n');
        }
        current.body.push_str(line);
    }
    flush(&mut current, &mut sections);
    if sections.is_empty() && !text.trim().is_empty() {
        return vec![Section {
            heading: String::new(),
            body: text.trim().to_owned(),
        }];
    }
    sections
}

fn match_heading(line: &str) -> Option<String> {
    let trimmed = line.trim().trim_matches(['：', ':']);
    let normalized = trimmed.to_lowercase();
    SECTION_HEADINGS
        .iter()
        .find(|heading| normalized == heading.to_lowercase())
        .map(|heading| (*heading).to_owned())
}

fn split_section(candidate: &str, heading: &str, body: &str) -> Vec<String> {
    if is_experience_heading(heading) {
        return chunk_blocks(candidate, heading, &split_experience_blocks(body));
    }
    if heading.is_empty() {
        let mut blocks = split_blank_paragraphs(body);
        if blocks.len() <= 1 {
            let items = split_bullet_items(body);
            if items.len() > 1 {
                blocks = items;
            }
        }
        return chunk_blocks(candidate, heading, &blocks);
    }
    let title = first_line_title(body);
    split_oversized(&with_prefix(candidate, heading, &title, body))
}

fn chunk_blocks(candidate: &str, heading: &str, blocks: &[String]) -> Vec<String> {
    let mut out = Vec::new();
    for block in blocks {
        let title = first_line_title(block);
        out.extend(split_oversized(&with_prefix(
            candidate, heading, &title, block,
        )));
    }
    out
}

fn split_blank_paragraphs(body: &str) -> Vec<String> {
    let mut blocks = Vec::new();
    let mut current = Vec::new();
    let flush = |current: &mut Vec<&str>, blocks: &mut Vec<String>| {
        let text = current.join("\n").trim().to_owned();
        if !text.is_empty() {
            blocks.push(text);
        }
        current.clear();
    };
    for line in body.lines() {
        if line.trim().is_empty() {
            flush(&mut current, &mut blocks);
            continue;
        }
        current.push(line);
    }
    flush(&mut current, &mut blocks);
    if blocks.is_empty() && !body.trim().is_empty() {
        return vec![body.trim().to_owned()];
    }
    blocks
}

fn is_experience_heading(heading: &str) -> bool {
    matches!(
        heading.to_lowercase().as_str(),
        "工作经历"
            | "工作经验"
            | "实习经历"
            | "项目经历"
            | "项目经验"
            | "experience"
            | "work experience"
            | "projects"
    )
}

fn split_experience_blocks(body: &str) -> Vec<String> {
    let mut blocks = Vec::new();
    let mut current = Vec::new();
    let flush = |current: &mut Vec<&str>, blocks: &mut Vec<String>| {
        let text = current.join("\n").trim().to_owned();
        if !text.is_empty() {
            blocks.push(text);
        }
        current.clear();
    };
    for line in body.lines() {
        if is_date_line(line.trim()) && !current.is_empty() {
            flush(&mut current, &mut blocks);
        }
        current.push(line);
    }
    flush(&mut current, &mut blocks);
    if blocks.is_empty() && !body.trim().is_empty() {
        return vec![body.trim().to_owned()];
    }
    blocks
}

fn first_line_title(block: &str) -> String {
    let line = block.lines().next().unwrap_or("").trim();
    let runes: Vec<char> = line.chars().collect();
    if runes.len() > 80 {
        runes[..80].iter().collect()
    } else {
        line.to_owned()
    }
}

fn with_prefix(candidate: &str, heading: &str, title: &str, body: &str) -> String {
    let label = [candidate, heading, title]
        .into_iter()
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>()
        .join(" | ");
    if label.is_empty() {
        return body.trim().to_owned();
    }
    format!("[{label}]\n{}", body.trim())
}

fn split_oversized(text: &str) -> Vec<String> {
    if text.chars().count() <= MAX_CHUNK_RUNES {
        if text.trim().is_empty() {
            return Vec::new();
        }
        return vec![text.to_owned()];
    }
    let (prefix, body) = match text.split_once("]\n") {
        Some((head, rest)) if text.starts_with('[') => (format!("{head}]\n"), rest.to_owned()),
        _ => (String::new(), text.to_owned()),
    };
    let items = split_bullet_items(&body);
    if items.len() > 1 {
        return items
            .into_iter()
            .flat_map(|item| split_oversized(&format!("{prefix}{}", item.trim())))
            .collect();
    }
    hard_split(text, &prefix)
}

fn split_bullet_items(body: &str) -> Vec<String> {
    let mut items = Vec::new();
    let mut current = Vec::new();
    let mut found = false;
    let flush = |current: &mut Vec<&str>, items: &mut Vec<String>| {
        let text = current.join("\n").trim().to_owned();
        if !text.is_empty() {
            items.push(text);
        }
        current.clear();
    };
    for line in body.lines() {
        if is_bullet_line(line.trim()) && !current.is_empty() {
            found = true;
            flush(&mut current, &mut items);
        }
        current.push(line);
    }
    flush(&mut current, &mut items);
    if !found {
        return Vec::new();
    }
    items
}

fn hard_split(text: &str, prefix: &str) -> Vec<String> {
    let runes: Vec<char> = text.chars().collect();
    let mut out = Vec::new();
    let mut start = 0;
    while start < runes.len() {
        let remaining = &runes[start..];
        if remaining.len() <= MAX_CHUNK_RUNES {
            let mut piece = remaining.iter().collect::<String>();
            piece = piece.trim().to_owned();
            if start > 0 && !prefix.is_empty() && !piece.starts_with('[') {
                piece = format!("{prefix}{piece}").trim().to_owned();
            }
            if !piece.is_empty() {
                out.push(piece);
            }
            break;
        }
        let window = &remaining[..MAX_CHUNK_RUNES];
        let mut cut = find_split(window);
        if cut == 0 {
            cut = MAX_CHUNK_RUNES;
        }
        let mut piece = remaining[..cut].iter().collect::<String>();
        piece = piece.trim().to_owned();
        if start > 0 && !prefix.is_empty() && !piece.starts_with('[') {
            piece = format!("{prefix}{piece}").trim().to_owned();
        }
        if !piece.is_empty() {
            out.push(piece);
        }
        start += cut;
    }
    out
}

fn find_split(window: &[char]) -> usize {
    if window.is_empty() {
        return 0;
    }
    let begin = window.len().saturating_sub(HARD_SPLIT_LOOKBACK);
    for i in (begin..window.len()).rev() {
        if window[i] == '\n' {
            if i + 1 < window.len() && window[i + 1] == '\n' {
                return i + 2;
            }
            return i + 1;
        }
    }
    for i in (begin..window.len()).rev() {
        if window[i] == '。' || window[i] == '！' || window[i] == '？' {
            if i > 0 && window[i - 1] == window[i] {
                continue;
            }
            return i + 1;
        }
    }
    0
}

fn locate_offsets(haystack: &str, content: &str) -> (usize, usize) {
    let needle = if content.starts_with('[') {
        content
            .split_once("]\n")
            .map(|(_, body)| body)
            .unwrap_or(content)
    } else {
        content
    };
    if let Some(byte) = haystack.find(needle) {
        let start = haystack[..byte].chars().count();
        return (start, start + needle.chars().count());
    }
    (0, content.chars().count())
}

fn is_date_line(line: &str) -> bool {
    let bytes = line.as_bytes();
    if bytes.len() < 6 {
        return false;
    }
    let mut index = 0;
    if !take_digits(line, &mut index, 4, 4) {
        return false;
    }
    let Some(sep) = line[index..].chars().next() else {
        return false;
    };
    if !matches!(sep, '.' | '/' | '-' | '年') {
        return false;
    }
    index += sep.len_utf8();
    if !take_digits(line, &mut index, 1, 2) {
        return false;
    }
    if line[index..].starts_with('月') {
        index += '月'.len_utf8();
    }
    let rest = line[index..].trim_start();
    if rest.is_empty() {
        return true;
    }
    let rest = match rest
        .strip_prefix("–")
        .or_else(|| rest.strip_prefix("—"))
        .or_else(|| rest.strip_prefix('-'))
        .or_else(|| rest.strip_prefix("至"))
        .or_else(|| rest.strip_prefix("到"))
    {
        Some(rest) => rest.trim_start(),
        None => return true,
    };
    let lowered = rest.to_ascii_lowercase();
    if lowered.starts_with("至今") || lowered.starts_with("现在") || lowered.starts_with("present")
    {
        return true;
    }
    let mut end = 0;
    take_digits(rest, &mut end, 4, 4)
        && rest[end..]
            .chars()
            .next()
            .is_some_and(|sep| matches!(sep, '.' | '/' | '-' | '年'))
}

fn take_digits(text: &str, index: &mut usize, min: usize, max: usize) -> bool {
    let mut count = 0;
    let bytes = text.as_bytes();
    while *index < bytes.len() && bytes[*index].is_ascii_digit() && count < max {
        *index += 1;
        count += 1;
    }
    count >= min
}

fn is_bullet_line(line: &str) -> bool {
    let Some(first) = line.chars().next() else {
        return false;
    };
    if matches!(first, '•' | '·' | '●' | '○' | '■' | '-' | '*' | '、') {
        return line[first.len_utf8()..].starts_with(|ch: char| ch.is_whitespace());
    }
    if first == '（' || first == '(' {
        let close = if first == '（' { '）' } else { ')' };
        let rest = &line[first.len_utf8()..];
        let digits = rest.chars().take_while(|ch| ch.is_ascii_digit()).count();
        if digits == 0 {
            return false;
        }
        return rest[digits..].starts_with(close);
    }
    let digits = line.chars().take_while(|ch| ch.is_ascii_digit()).count();
    if digits == 0 {
        return false;
    }
    line[digits..].starts_with(['.', '、'])
}

#[cfg(test)]
mod tests {
    use super::{CHUNKER_VERSION, chunk_text};

    #[test]
    fn chunker_version_is_stable() {
        assert_eq!(CHUNKER_VERSION, "resume-semantic-v1");
    }

    #[test]
    fn keeps_whole_job_block() {
        let text = "工作经历\n2019.03-2021.06 阿里巴巴 高级工程师\n负责订单服务与 Kafka 链路，使用 Node.js 完成 2019.03 版本升级。\n日常维护 Vue.js 与 v1.2 配置，数值约 3.14。";
        let chunks = chunk_text("张三", text);
        assert_eq!(chunks.len(), 1);
        assert!(chunks[0].content.contains("Node.js"));
        assert!(chunks[0].content.contains("2019.03 版本"));
        assert!(
            chunks[0]
                .content
                .starts_with("[张三 | 工作经历 | 2019.03-2021.06 阿里巴巴 高级工程师]")
        );
        assert_eq!(chunks[0].section, "工作经历");
    }

    #[test]
    fn splits_on_date_lines_only() {
        let text = "工作经历\n2019.03-2021.06 阿里巴巴\n负责 Node.js 与 2019.03 内部版本。\n2021.07-至今 腾讯\n负责 Kafka 订单链路。";
        let chunks = chunk_text("李四", text);
        assert_eq!(chunks.len(), 2);
        assert!(chunks[0].content.contains("阿里巴巴"));
        assert!(!chunks[0].content.contains("腾讯"));
        assert!(chunks[1].content.contains("腾讯"));
    }

    #[test]
    fn splits_oversized_by_bullets_with_repeated_prefix() {
        let body = "职责说明".repeat(280);
        let text = format!("工作经历\n2018.01-2019.01 某公司\n• {body}\n• {body}");
        let chunks = chunk_text("王五", &text);
        assert!(
            chunks.len() >= 2,
            "expected bullet split, got {}",
            chunks.len()
        );
        for chunk in &chunks {
            assert!(
                chunk.content.chars().count() <= 2000,
                "chunk exceeded safety valve: {}",
                chunk.content.chars().count()
            );
            assert!(
                chunk
                    .content
                    .contains("[王五 | 工作经历 | 2018.01-2019.01 某公司]")
            );
        }
    }

    #[test]
    fn without_headings_uses_blank_lines() {
        let chunks = chunk_text("赵六", "第一段公司经历内容。\n\n第二段项目经历内容。");
        assert_eq!(chunks.len(), 2);
        let joined = chunks
            .iter()
            .map(|chunk| chunk.content.as_str())
            .collect::<Vec<_>>()
            .join("\n");
        assert!(!joined.contains(&"第一段".repeat(2)));
    }
}
